### PR TITLE
Fix filters context propagation + ordering

### LINKED ISSUE #5400: Regression ServerRequestContext::currentRequest not working in HTTP Filters since 2.5.0
`ServerRequestContext.currentRequest();`  is not resolving the request since 2.5.0 in a Http Server Filter. 

Given the following controllers: 

```java
package example.micronaut;

import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;

@Controller("/es")
public class EsController {

    @Produces(MediaType.TEXT_HTML)
    @Get
    public String index() {
        return "es";
    }
}
```

```java
package example.micronaut;

import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;

@Controller("/en")
public class EnController {

    @Produces(MediaType.TEXT_HTML)
    @Get
    public String index() {
        return "en";
    }
}
```

```java
package example.micronaut;

import io.micronaut.core.async.publisher.Publishers;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.annotation.Filter;
import io.micronaut.http.context.ServerRequestContext;
import io.micronaut.http.filter.OncePerRequestHttpServerFilter;
import io.micronaut.http.filter.ServerFilterChain;
import org.reactivestreams.Publisher;
import java.net.URI;
import java.util.Optional;

@Filter(Filter.MATCH_ALL_PATTERN)
public class LangFilter extends OncePerRequestHttpServerFilter {
    @Override
    protected Publisher<MutableHttpResponse<?>> doFilterOnce(HttpRequest<?> request, ServerFilterChain chain) {
        if (request.getUri().toString().equals("/es")) {
            return chain.proceed(request);
        }
        Optional<HttpRequest<Object>> optionalHttpRequest = ServerRequestContext.currentRequest();
        if (optionalHttpRequest.isPresent()) {
            String accepts = optionalHttpRequest.get().getHeaders().get("Accept-Language");
            if (accepts != null && accepts.equals("es")) {
                return Publishers.just(HttpResponse.seeOther(URI.create("/es")));
            }
        }
        return chain.proceed(request);
    }
}
```

With 2.4.4, I am able to get hold of the current request with `ServerRequestContext`

```
curl -iL -H "Accept-Language: es" localhost:8080/en
HTTP/1.1 303 See Other
Location: /es
Date: Sat, 8 May 2021 02:47:15 GMT
connection: keep-alive
transfer-encoding: chunked

HTTP/1.1 200 OK
Date: Sat, 8 May 2021 02:47:15 GMT
content-type: text/html
content-length: 2
connection: keep-alive

es
```

with 2.5.0 or 2.5.1, it always returns optional: 

```
curl -iL -H "Accept-Language: es" localhost:8080/en
HTTP/1.1 200 OK
Date: Sat, 8 May 2021 02:50:05 GMT
content-type: text/html
content-length: 2
connection: keep-alive

en
```

### PR DESCRIPTION
**Pull request by dstepanov**
Fixed #5400

### CODE REVIEW

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java [hunk 1]
```diff
@@ -2143,8 +2143,18 @@ private ByteBuf encodeBodyAsByteBuf(Object body, MediaTypeCodec codec, ChannelHa
...
+            ServerRequestContext.set(requestReference.get());
+            HttpFilter httpFilter = filters.get(0);
+            return (Publisher<MutableHttpResponse<?>>) httpFilter.doFilter(requestReference.get(), filterChain);
+        } finally {
+            if (prevRequest.isPresent()) {
+                ServerRequestContext.set(prevRequest.get());
+            } else {
+                ServerRequestContext.set(null);
+            }
+        }
```
Comment (graemerocher) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java:2157]: Could you not use `ServerRequestContext.with(requestReference.get(), () -> )`?

Comment (dstepanov) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java:2157]: I can but I don't know if it's worse to have a new lambda always being created there?

Comment (graemerocher) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java:2157]: Perhaps

**Review by graemerocher — APPROVED** [review]

**Comment by jameskleeh** [comment]
Closing in favor of https://github.com/micronaut-projects/micronaut-core/pull/5413

**Pull request closed without merging**